### PR TITLE
WFLY-6944 Encode session affinity using multiple routes, if supported by the load balancer

### PR DESCRIPTION
**Pull request by pferraro**
https://issues.jboss.org/browse/WFLY-6944

Analysis doc: https://github.com/wildfly/wildfly-proposals/pull/150

Requires https://github.com/wildfly/wildfly/pull/11476

### CODE REVIEW

**Comment by bstansberry** [comment]
retest this please

**Comment by bstansberry** [comment]
The 2 commits in this PR that go beyond what's in #11476 LGTM.

File: clustering/web/extension/pom.xml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>org.wildfly</groupId>
+        <artifactId>wildfly-clustering-web</artifactId>
+        <!--
+        Maintain separation between the artifact id and the version to help prevent
+        merge conflicts between commits changing the GA and those changing the V.
+        -->
+        <version>16.0.0.CR1-SNAPSHOT</version>
```
Comment (rhusar) [inline clustering/web/extension/pom.xml]: Please rebase with:
```
-         <version>16.0.0.CR1-SNAPSHOT</version>
+         <version>17.0.0.Beta1-SNAPSHOT</version>
```

File: clustering/web/container/pom.xml [hunk 2]
```diff
@@ -20,11 +20,31 @@
...
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>org.wildfly</groupId>
+        <artifactId>wildfly-clustering-web</artifactId>
+        <!--
+        Maintain separation between the artifact id and the version to help prevent
+        merge conflicts between commits changing the GA and those changing the V.
+        -->
+        <version>16.0.0.CR1-SNAPSHOT</version>
```
Comment (rhusar) [inline clustering/web/container/pom.xml]: Please rebase with:
```
-         <version>16.0.0.CR1-SNAPSHOT</version>
+         <version>17.0.0.Beta1-SNAPSHOT</version>
```

Comment (pferraro) [reply clustering/web/container/pom.xml]: This file does not originate with this PR.

File: clustering/web/extension/pom.xml (hunk 1 above)
Comment (pferraro) [reply clustering/web/extension/pom.xml]: This file does not originate with this PR.

File: clustering/web/extension/src/main/resources/schema/distributable-web_1_0.xsd [hunk 3]
```diff
@@ -106,6 +115,19 @@
...
 
+    <xs:complexType name="ranked-affinity">
+        <xs:attribute name="delimiter" type="xs:string" default=".">
+            <xs:annotation>
+                <xs:documentation>The delimiter used to separate ranked routes within the session ID.</xs:documentation>
+            </xs:annotation>
+        </xs:attribute>
+        <xs:attribute name="max-routes" type="xs:string" default="3">
+            <xs:annotation>
+                <xs:documentation>The maximun number of routes to append to the session ID.</xs:documentation>
```
Comment (rhusar) [inline clustering/web/extension/src/main/resources/schema/distributable-web_1_0.xsd]: maximun*

File: clustering/web/extension/src/main/resources/org/wildfly/extension/clustering/web/LocalDescriptions.properties [hunk 4]
```diff
@@ -36,3 +36,6 @@ distributable-web.affinity.remove=Removes an affinity configuration
 distributable-web.affinity.none=Web requests will not have an affinity for any cluster member
 distributable-web.affinity.local=Web requests will have an affinity for the member that last handled a given session
 distributable-web.affinity.primary-owner=Web requests will have an affinity for the primary owner of a given session
+distributable-web.affinity.ranked=Web requests will have an affinity for the first available node in a list containing primary and backup owners, and the member that last handled a given
```
Comment (tommaso-borgato) [inline clustering/web/extension/src/main/resources/org/wildfly/extension/clustering/web/LocalDescriptions.properties]: is the world "session" needed at the end of the phrase?

**Comment by rhusar** [comment]
@pferraro Needs a rebase :(

**Comment by pferraro** [comment]
@rhusar It's worse than that - I have to increment schema, model, and deployment schema versions since this will target WF18...

**Comment by rhusar** [comment]
*sigh*

File: clustering/web/infinispan/src/test/java/org/wildfly/clustering/web/infinispan/routing/RankedRouteLocatorTestCase.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                Assert.assertEquals("2.0.1", result);
+                break;
+            }
+            default: {
+                String result = locator.locate("session");
+                Assert.assertEquals("local", result);
+            }
+        }
+    }
+}
```
Comment (tommaso-borgato) [inline clustering/web/infinispan/src/test/java/org/wildfly/clustering/web/infinispan/routing/RankedRouteLocatorTestCase.java]: @pferraro could you give me a quick explanation about this test? I have difficulties understanding what it exactly does

Comment (pferraro) [reply clustering/web/infinispan/src/test/java/org/wildfly/clustering/web/infinispan/routing/RankedRouteLocatorTestCase.java]: @tommaso-borgato It validates that the RouteLocator implementation returns the expected results from a given cache topology for a given cache mode.

Comment (tommaso-borgato) [reply clustering/web/infinispan/src/test/java/org/wildfly/clustering/web/infinispan/routing/RankedRouteLocatorTestCase.java]: @pferraro is it correct to say that it tests that `JVM route` portion of the JSESSIONID cookie is correctly filled in based on the cache topology?

Comment (pferraro) [reply clustering/web/infinispan/src/test/java/org/wildfly/clustering/web/infinispan/routing/RankedRouteLocatorTestCase.java]: Yes.

File: clustering/web/extension/src/main/resources/org/wildfly/extension/clustering/web/LocalDescriptions.properties (hunk 4 above)
Comment (pferraro) [reply clustering/web/extension/src/main/resources/org/wildfly/extension/clustering/web/LocalDescriptions.properties]: Yes indeed.

**Comment by bstansberry** [comment]
@rhusar Look ok to you? Seems ok to me but you had comments earlier plus a 2nd approval from someone more accustomed to the clustering style of management would be good.

Ping me if ok, and I'll merge this. Assuming fresh ci run i'll start now is still ok.

**Comment by bstansberry** [comment]
retest this please

**Pull request merged**